### DISCUSSION TITLE
Extend Zodios Instance

### DISCUSSION THREAD

**Discussion by stychu** [ecyrbe/zodios#213] [Q&A] [answered]
Hello there, thanks for Zodios! I've been looking for good typesafe http client builder and Zodios seems the one ! <3
I'm still a newbie with TypeScript and wanted to know how I could extend ZodiosInstance type with custom function of mine.

I need to setup Bearer token on demand when request is coming to the server but I don't like code dangling around if not necessary. So I would like to wrap everything Zodios setup related in my index file.

How I can extend my Zodios client with such function 

```js
const setAuthorizationHeader = (value: string) =>
  tlsApiClient.use(pluginToken({ getToken: async () => value }));

```

```js
import { Zodios } from "@zodios/core";
import { env } from "@/env/server.mjs";
import { pluginHeader } from "@zodios/plugins";
import { userApi } from "@/server/services/tlsApi/user";

const tlsApiClient = new Zodios(env.LOCATION_SERVICE_BASE_URL, [...userApi]);

tlsApiClient.use(
  pluginHeader("User-Agent", async () => env.LOCATION_SERVICE_USER_AGENT)
);

export type TlsApi = typeof tlsApiClient;

export { tlsApiClient };

```

Ofc I can just export that function and use it but I would prefer do something like this whenever I use the client
```js
tlsApiClient.setAuthorizationHeader('asd')
```

**Comment by ecyrbe** [comment, marked as answer]
You can do this:
```ts
class ZodiosExtended extends Zodios<typeof api> {
  constructor(myAPI: typeof api) {
    super(myAPI);
  }
  setAuthorisationHeader(token: string) {
    this.use(pluginToken({ getToken: async () => token }));
  }
}
```

This will be an instance specific to one API though (not generic). 
Indeed, typescript/javascript can't let you extends generic classes, it needs to know what it extends at declaration time.

If you still want to have generics, you'll need to do mixins and and do the same thing zodios does to allow alias definition:
you can take a look at [how zodios does it](https://github.com/ecyrbe/zodios/blob/main/src/zodios.ts#L414), but i strongly encourage you to not do it, it's a lot more work than having a free form function for litle value.

**Reply by thelinuxlich** [reply]
You can also pass it to axios options inside Zodios constructor, right?

**Reply by ecyrbe** [reply]
Hello @thelinuxlich, 
It will not exactly address the question :
- zodios plugins are not axios interceptors, so you'll need to write an axios interceptor
- @stychu wants to extend zodios to add it own methods, passing an axios instance will still not address how to extend zodios

**Reply by thelinuxlich** [reply]
I thought it was restricted to adding headers on setup

**Reply by ecyrbe** [reply]
Usually, bearer token do change at runtime, so you want to use an interceptor or a plugin for that. 
But yes, if the headers you want to inject are known at instanciation time and don't change later, you can use a default axios config on zodios constructor.

**Reply by thelinuxlich** [reply]
true, wasn't thinking clearly, needs more coffee

**Reply by stychu** [reply]
Thanks @ecyrbe for suggestion! I went with simple approach to not overcomplicate things ;) 

```
const tlsApi = new Zodios(env.LOCATION_SERVICE_BASE_URL, [...userApi]);

tlsApi.use(pluginUnauthorized());
tlsApi.use(pluginHeader("X-Authorization-Type", async () => "onelogin"));
tlsApi.use(
  pluginHeader("User-Agent", async () => env.LOCATION_SERVICE_USER_AGENT)
);

function setAuthorizationHeader(token: string) {
  tlsApi.use(pluginToken({ getToken: async () => token }));
}

const tlsApiClient = {
  tlsApi,
  setAuthorizationHeader,
};
```
